### ISSUE TITLE
Use http://electron.atom.io to make standalone applications

### ISSUE THREAD

**Issue by jeffreyrosenbluth** [HeinrichApfelmus/threepenny-gui#111]

**Comment by HeinrichApfelmus** [comment]
I have already taken a look at Electron and very much liked what I saw. I think this is the most promising avenue for standalone applications with JavaScript.

**Comment by jeffreyrosenbluth** [comment]
That's great, let me know if you would like any help.

**Comment by HeinrichApfelmus** [comment]
@jeffreyrosenbluth Since you're offering help: Writing JS driver code the Electron shell should be no problem for me, but I wonder how or if it's possible to integrate this with cabal. It would be great if we could build some kind of `cabal-electron` package that offers post-build hooks to assemble package bundles from Haskell source files. I'm thinking of something similar to [`cabal-macosx`][1].

  [1]: https://github.com/gimbo/cabal-macosx

**Comment by jeffreyrosenbluth** [comment]
Let me see what would be involved and whether is seems do-able.

**Comment by jeffreyrosenbluth** [comment]
This doesn't seem like it would be too difficult so let me see if I understand correctly what would be required. First, if we were going to do it "by hand" we would build the user's application, say app.hs, which imports the threepenny modules to a binary using cabal build. Then we would package this binary together with the JS driver code, a package.json file, and an index.html file if desired we could then use an asar archive to make this into a single app file. The package.json and index.html files would be the same for every app?

If this is correct I think that cabal integration would be fairly straight forward. Much of cabal-macosx deals with handling dependencies. I don't think we would have any dependencies that are not handled by regular cabal or are included in our static js code.

Let me know if you agree with this assessment and we can decide the best way to proceed.

**Comment by HeinrichApfelmus** [comment]
Yes, pretty much. Essentially, we need a post-build hook that copies the executable and the index.html and package.json files into a common directory, so that Electron can find a complete directory structure.

My problem is that I don't know much about Cabal. If you could figure out how to make such a post-build hook, then that would help me a lot.

**Comment by jeffreyrosenbluth** [comment]
I don't know that much about Cabal either but I think I'll be able to figure it out.
I'll start working on it.

**Comment by HeinrichApfelmus** [comment]
I have successfully hooked up Threepenny-GUI to Electron.

I still don't know how to offer this functionality via .cabal files, though. Electron is an external dependency, the precompiled binary weighs 30-40MB (for [version 0.33][1]) for each platform. Application bundles have to be built by copying the relevant files.

  [1]: https://github.com/atom/electron/releases/tag/v0.33.6

**Comment by jeffreyrosenbluth** [comment]
I have to apologize, I thought I would have time to work on this, but I'm tied up on a large project for the foreseeable future. Sorry about that, I may still be able to do it, just not for a while.

**Comment by HeinrichApfelmus** [comment]
No worries, this is a completely voluntary activity, after all. :smile:

**Comment by tr00per** [comment]
> I have successfully hooked up Threepenny-GUI to Electron.
Do you consider releasing a HOWTO for that?

**Comment by bradrn** [comment]
I think I've figured out how to hook up Threepenny-GUI to Electron. What you do is:
- Build the Haskell program and move the executable to the root of the program
- Add this as your `package.json`:
  ```json
  {
      "name": "<name>",
      "version": "0.1.0",
      "main": "./main.js",
      "scripts": {
          "start": "electron ."
      }
  }
  ```
- Run `npm install --save-dev electron`
- Put this as your `index.html`:
  ```html
  <webview src="http://localhost:<port>" style="width: 100%; height: 100%"></webview>
  ```
- Add the quickstart `main.js` file [here](https://github.com/electron/electron-quick-start/blob/master/main.js).
- Add the following to `main.js` just before it says `mainWindow.loadURL(url.format({` (line 19):
  ```javascript
  const execFile = require('child_process').execFile;
  const child = execFile('./<haskell-program-name>');
  ```
- You might want to remove the lines of code which open the developer tools from `main.js` (it's well-commented, so they should be easy to find)
- Run the program using `npm start`
I've managed to get the sample programs distributed with threepenny-gui working using this method.

**Comment by barischj** [comment]
Here's method of doing this without custom HTML, it starts the app on a **random port** and **waits until the Threepenny app is running** before starting Electron.

- Add this [package.json](https://gist.github.com/barischj/453ab805e0f0301e6a27433011bd5b1b) to root directory or add the three dependencies to your own `package.json`
- Add this [electron.js](https://gist.github.com/barischj/89ecd3ac6f7e556f8ce6e02a03cae3b9) to root directory
- Edit your Haskell `main` to take a port as an argument and run on that port, something like:
``` Haskell
module Main where

import System.Environment  (getArgs)
import YourApp (start)

main :: IO ()
main = do
    [port] <- getArgs
    start (read port)
```
- Edit line 41 of `electron.js` to run **your** application, currently:
``` Node
child = spawn('stack', ['exec', 'your-app', `${port}`]);
```

Now to run the whole thing:
``` Bash
npm install
stack build
./node_modules/.bin/electron electron.js
```

**Comment by bradrn** [comment]
> ... without moving the executable ...

The reason I used the method of 'moving the executable' instead of `stack exec <program>` is that the threepenny-gui executable can then be used on a computer without `stack`.

---

I'm going to create a repo with all the config files in it as an example of how to integrate threepenny-gui and electron. I'm having a bit of trouble with pushing the files though so I'll do it after I fix that.
